File: javascript/packages/client/test/names.test.ts
import { describe, test, expect, beforeEach } from "vitest"
import { SlotIndex } from "../src/slot-index"

const FILE = "app/views/posts/index.html.erb"

const PAGE =
  `<!--herb-region:${FILE}:aaaaaaaa:0-->` +
  `<ul data-herb-name="0:messages"><!--herb-slot:0:collection-->` +
  `<!--herb-item:0:a--><li id="a" data-herb-slot="1:attribute:id"><span data-herb-name="2:body" data-herb-slot="2:child">first</span></li><!--/herb-item:0-->` +
  `<!--herb-item:0:b--><li id="b" data-herb-slot="1:attribute:id"><span data-herb-name="2:body" data-herb-slot="2:child">second</span></li><!--/herb-item:0-->` +
  `<!--/herb-slot:0--></ul>` +
  `<p data-herb-name="3:summary"><!--herb-slot:3-->two messages<!--/herb-slot:3--></p>` +
  `<!--/herb-region:${FILE}-->`

let index: SlotIndex

beforeEach(() => {
  document.body.innerHTML = PAGE

  index = new SlotIndex()
  index.scan(document.body)
})

describe("addressing a slot by name", () => {
  test("resolves a region-level name to its slot", () => {
    const slot = index.slot(FILE, "summary")

    expect(slot).not.toBeNull()
    expect(slot!.index).toBe(3)
  })

  test("resolves a named collection", () => {
    const slot = index.slot(FILE, "messages")

    expect(slot?.type).toBe("collection")
    expect(index.itemsFor(FILE, "messages").size).toBe(2)
  })

  test("resolves an item's slot per row through the item key", () => {
    const first = index.slotInItem(FILE, "messages", "a", "body")
    const second = index.slotInItem(FILE, "messages", "b", "body")

    expect(index.currentText(first!)).toBe("first")
    expect(index.currentText(second!)).toBe("second")
    expect(first).not.toBe(second)
  })

  test("resolves an attribute slot by its attribute name with nothing authored", () => {
    const slot = index.slotInItem(FILE, "messages", "a", "id")

    expect(slot?.attribute).toBe("id")
  })

  test("answers null for a name nothing declares", () => {
    expect(index.slot(FILE, "missing")).toBeNull()
    expect(index.slotInItem(FILE, "messages", "a", "missing")).toBeNull()
  })

  test("an index keeps working everywhere a name does", () => {
    expect(index.slot(FILE, "summary")).toBe(index.slot(FILE, 3))
    expect(index.slotInItem(FILE, 0, "a", 2)).toBe(index.slotInItem(FILE, "messages", "a", "body"))
  })
})

File: test/engine/slots/names_test.rb
# frozen_string_literal: true

require_relative "../../test_helper"
require_relative "../../snapshot_utils"
require_relative "../../../lib/herb/engine"
require_relative "../../../lib/herb/engine/slots/visitor"

module Engine
  module Slots
    class NamesTest < Minitest::Spec
      include SnapshotUtils

      def options
        { visitors: [Herb::Engine::Slots::Visitor.new(mode: :client)], filename: "app/views/test.html.erb" }
      end

      def compile(template)
        visitor = Herb::Engine::Slots::Visitor.new(mode: :client)
        engine = Herb::Engine.new(template, visitors: [visitor], filename: "app/views/test.html.erb")

        [visitor, engine.src]
      end

      def slots(template)
        compile(template).first.slots
      end

      test "records a name against the slot the element holds" do
        named = slots(%(<p data-herb-name="body"><%= @body %></p>)).find { |slot| slot.name == "body" }

        assert named
        assert_equal :child, named.type
      end

      test "leaves the name the template wrote where the template wrote it" do
        assert_evaluated_snapshot(%(<p data-herb-name="body"><%= @body %></p>), { "@body" => "hi" }, options)
      end

      test "says which slot that name is, in the manifest" do
        visitor, = compile(%(<p data-herb-name="body"><%= @body %></p>))
        index = visitor.slots.find { |slot| slot.name == "body" }.index

        assert_equal({ "body" => index }, visitor.manifest["names"])
      end

      test "mixed static text around one slot still resolves" do
        named = slots(%(<p data-herb-name="greeting">Hello <%= @name %>!</p>)).find { |slot| slot.name == "greeting" }

        assert named
        assert_equal :child, named.type
      end

      test "names the collection an element holds" do
        template = %(<ul data-herb-name="messages"><% @messages.each do |m| %><li id="<%= m %>"><%= m %></li><% end %></ul>)
        named = slots(template).find { |slot| slot.name == "messages" }

        assert named
        assert_equal :collection, named.type
      end

      test "an element with two content slots is ambiguous" do
        error = assert_raises(Herb::Engine::CompilationError) do
          compile(%(<p data-herb-name="pair"><%= @first %> and <%= @second %></p>))
        end

        assert_equal "`data-herb-name=\"pair\"` on `<p>` is ambiguous between 2 slots. Wrap the one it should name in its own element.", error.message
      end

      test "an element holding nothing dynamic names no slot" do
        error = assert_raises(Herb::Engine::CompilationError) do
          compile(%(<p data-herb-name="static">plain text</p>))
        end

        assert_equal "`data-herb-name=\"static\"` on `<p>` names no slot, since the element holds nothing dynamic. Move the name onto an element that wraps an ERB output, or remove it.", error.message
      end

      test "a computed name is refused" do
        error = assert_raises(Herb::Engine::CompilationError) do
          compile(%(<p data-herb-name="<%= @name %>"><%= @body %></p>))
        end

        assert_equal "`data-herb-name` on `<p>` is computed or empty. A slot name is an address the browser looks up, so give it a static, non-empty value.", error.message
      end

      test "two slots in one scope cannot share a name" do
        error = assert_raises(Herb::Engine::CompilationError) do
          compile(%(<p data-herb-name="body"><%= @a %></p><div data-herb-name="body"><%= @b %></div>))
        end

        assert_equal "Two slots in the same scope are both named `body`. A slot name is an address, so give one of them a different name.", error.message
      end

      test "a name may repeat between an item and its region" do
        template = <<~ERB
          <p data-herb-name="body"><%= @intro %></p>
          <ul><% @items.each do |item| %><li id="<%= item %>"><span data-herb-name="body"><%= item %></span></li><% end %></ul>
        ERB

        named = slots(template).select { |slot| slot.name == "body" }

        assert_equal 2, named.size
      end

      test "a name colliding with an attribute slot in the same scope is refused" do
        error = assert_raises(Herb::Engine::CompilationError) do
          compile(%(<div id="<%= @id %>"><p data-herb-name="id"><%= @body %></p></div>))
        end

        assert_equal "The name `id` collides with the `id` attribute slot in the same scope. An attribute slot is already addressable by its attribute, so drop the name or rename it.", error.message
      end

      test "a collapsed conditional does not shift the collision check" do
        template = <<~ERB
          <% if @a %><b><%= @p %></b><% else %><b><%= @q %></b><% end %>
          <% for item in @items %>
            <input sel="<%= item.id %>">
          <% end %>
          <div data-herb-name="sel"><%= @y %></div>
        ERB

        named = slots(template).find { |slot| slot.name == "sel" }

        assert named
        assert_equal :child, named.type
      end

      test "two named elements cannot claim one slot" do
        error = assert_raises(Herb::Engine::CompilationError) do
          compile(%(<div data-herb-name="outer"><span data-herb-name="inner"><%= @x %></span></div>))
        end

        assert_equal "`data-herb-name=\"outer\"` claims the slot already named `inner`. Both elements hold the same slot, so keep one name or wrap what each should address in its own element.", error.message
      end

      test "a nested container's slots do not confuse the count" do
        template = %(<div data-herb-name="status"><% if @ok %><b><%= @yes %></b><% else %><%= @no %><% end %></div>)
        named = slots(template).find { |slot| slot.name == "status" }

        assert named
        assert_equal :conditional, named.type
      end

      test "the name is part of the version" do
        plain = compile(%(<p><%= @body %></p>)).first.version
        named = compile(%(<p data-herb-name="body"><%= @body %></p>)).first.version

        refute_equal plain, named
      end
    end
  end
end
